### PR TITLE
expand EDQUOT variable

### PR DESCRIPTION
**Pull request by avg-I**
See #4443

### CODE REVIEW

File: tests/zfs-tests/tests/functional/quota/quota.kshlib [hunk 1]
```diff
@@ -57,7 +57,7 @@ function fill_quota
 	$FILE_WRITE -o create -f $MNTPT/$TESTFILE1 -b $BLOCK_SIZE \
 		 -c $write_size -d 0
 	zret=$?
-        [[ $zret -ne EDQUOT ]] && \
+        [[ $zret -ne $EDQUOT ]] && \
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/quota/quota.kshlib]: Exactly which quota tests does this fix and can we enable a few more which are currently disabled in the runfile?  If so let's do it.

Comment (avg-I) [inline tests/zfs-tests/tests/functional/quota/quota.kshlib]: For me this fixes `quota_001_pos` and `quota_003_pos` on Gentoo.
I haven't tried any of the disabled test cases yet.

Comment (behlendorf) [inline tests/zfs-tests/tests/functional/quota/quota.kshlib]: OK, thanks for the clarification.

Comment (avg-I) [inline tests/zfs-tests/tests/functional/quota/quota.kshlib]: I've just enabled `quota_002_pos`, `quota_004_pos` and `quota_005_pos` and they passed for me too.
That's on Gentoo, I tried with `DISKS` pointing to real 40GB disks and with `DISKS` unset.

**Comment by behlendorf** [comment]
Interesting, those tests still fail for me on CentOS 7 with the default loopback devices.  But clearly it's something about the test case itself.

**Comment by behlendorf** [comment]
@avg-I and they're failing all for the same reason.  So we can either make this small improvement as is.  Or if you're feeling ambitious and want to fix the failure I'm seeing too we could then enable all these tests.

```
Test: /home/behlendo/src/git/zfs/tests/zfs-tests/tests/functional/quota/quota_002_pos (run as root) [00:00] [FAIL]
11:58:01.62 ASSERTION: Verify that a file write cannot exceed the file system quota
11:58:01.65 ERROR: /home/behlendo/src/git/zfs/cmd/zfs/zfs set quota=20000000 testpool.24272/testfs.24272 exited 1
11:58:01.66 cannot set property for 'testpool.24272/testfs.24272': size is less than current used or reserved space
11:58:01.66 NOTE: Performing local cleanup via log_onexit (cleanup)
```

**Comment by avg-I** [comment]
Interesting. Perhaps the loopback devices are simply too small?

**Comment by behlendorf** [comment]
@avg-I just the default 2G size.  I'm not sure why you wouldn't hit the same issue on Gentoo.

**Comment by avg-I** [comment]
@behlendorf argh, I forgot that I was using a modified version of ZFS when I ran those test.
So, this could be a regression (or an improvement?) in our local changes.
I'll retest with vanilla ZoL. Sorry about that.

**Comment by behlendorf** [comment]
@avg-I where does this stand?  Would you like to just make the small EQUOT fix for now or also enable (and fix) the additional disabled quota tests?

**Comment by avg-I** [comment]
I would prefer just the EDQUOT fix. I can not reproduce the failures mentioned in the run file _size is less than current used or reserved space_, so I am not sure if the EDQUOT fix has any effect on those disabled tests.

**Comment by avg-I** [comment]
FWIW, I think that I've found why I have this problem while nobody else does.
Gentoo:
```
$ ksh --version
  version         sh (AT&T Research) 93v- 2014-06-25
$ ksh
$ readonly EDQUOT=122
$ [[ 122 -eq EDQUOT ]] && echo true
ksh: EDQUOT: arithmetic syntax error
```

Centos 7:
```
$ ksh --version
  version         sh (AT&T Research) 93u+ 2012-08-01
$ ksh
$ readonly EDQUOT=122
$ [[ 122 -eq EDQUOT ]] && echo true
true
```

I have a newer ksh version on Gentoo, not sure if the difference in behavior is caused by a regression or by a bug fix.  In any case, explicitly expanding the variable should work for all versions.

**Comment by avg-I** [comment]
Also, I now have a theory about why those disabled tests don't fail for me.
I think that it is SELinux that might make a difference. I have it disabled on Gentoo, while it is enabled on CentOS.
My hypothesis is that a security context / label gets set on a temporary file and I believe that it is stored as an extended attribute. I further hypothesize that removing a file with extended attributes takes a different, longer path than removing a simpler file. So, my guess is that while the file gets removed, its space is not really returned until later. Thus, the cleanup done in, e.g., `quota_001_pos` does not really immediately return a filesystem to its original state and that does not allow to set the quota in `quota_002_pos`.

**Comment by behlendorf** [comment]
@avg-I that's fine, I don't mind merging the smaller fix first.  I just didn't want to merge something if you were still working on it.

As for the ksh version I don't know enough about ksh either to say what right correct behavior is here.  But as you say expanding the variable should, and does, work for both versions.

Regarding selinux that might be the issue.  If you look carefully at the `zfs-tests.sh` wrapper script you'll see it's already setting selinux to permissive mode so nothing is enforced.  However, it doesn't outright disable it as you've done so maybe that's the issue.  Are you able to reproduce the failures by using permissive mode again?

Another thing to be keenly aware of is that under Linux a `sync(1)` has _will not_ force a txg sync.  Several of the test cases make this assumption and exploit this side effect.  This doesn't work under Linux because by the time we get through the VFS to the ZFS specific code there's no way to tell between a sync and fsync.  So we always have to use `zfs_fsync()` never `zfs_sync()`.

For a long time I've wanted to add an explicit  `zfs sync` command for use by test cases but thus far it hasn't been a priority.

**Pull request closed without merging**